### DISCUSSION TITLE
[PROPOSAL]: Re-write with native TypeScript support

### DISCUSSION THREAD

**Discussion by michealroberts** [graphieros/vue-data-ui#112] [Ideas]
Firstly, looks like an awesome library. Some of the visualizations on the website are amazing. 🚀 Props to you @graphieros 

---

However, the dist bundled types are awkward to work with. They seem to be correct, however, not being able to, e.g.,:

```ts
import { type Config, VueUiXy } from "vue-data-ui";
```

Is leading to some tense situations with the TS language tools and compiler, e.g.: I've copied the base graph example from the builder here: https://vue-data-ui.graphieros.com/chart-builder

However, the types are awkward to derive:

<img width="1052" alt="Screenshot 2024-10-23 at 17 07 21" src="https://github.com/user-attachments/assets/639fdf5d-b0a0-4938-9e91-504aaa79f70a">

So, it leaves me in a quandry. Use a second rate library with first class TypeScript support, e.g., @unovis, or request that we discuss migrating the codebase to use .ts rather than a declaration file which is awkward to maintain and prone to breaking between minor and patch versions.

I am more than happy to offer up my services to help do this, as well as reconfigure the build process to output .d.ts and .cjs/.mjs/.js files.

Let me know your thoughts ...

**Comment by graphieros** [comment]
Hey @michealroberts :)

Before I answer, have you tried importing the proper types for this use case ?

```ts
import { VueUiXy, type VueUiXyConfig, type VueUiXyDatasetItem } from "vue-data-ui" 
```

**Comment by michealroberts** [comment]
Yes, I have also tried the above ... but certain values from the documentation are missing:

<img width="431" alt="Screenshot 2024-10-23 at 17 30 22" src="https://github.com/user-attachments/assets/b3ae07ce-2862-44f1-9797-143f88064e50">

**Comment by graphieros** [comment]
Yeah, that attribute is missing from the type def.
I would need to recheck the whole .d.ts file

But without this kind of mishap wouldn't it be sufficient ?

**Comment by michealroberts** [comment]
Yeh it's an awesome library ... I just think these mishaps happen because of the lack of explicit types close to the code ... the minute you have to maintain something external to where something is, the more problematic it will become ...

**Comment by michealroberts** [comment]
I'm more than happy to dedicate a day to this effort ...

**Comment by graphieros** [comment]
To check the .d.ts file, or to use lang="ts" on all components ? The latter is a massive undertaking

**Comment by graphieros** [comment]
@michealroberts listen I'm gonna start by cleaning up the .d.ts. There are quite a bunch of mishaps.
Maybe we can talk about your proposal through a different medium ?

**Comment by graphieros** [comment]
@michealroberts [v2.3.37](https://github.com/graphieros/vue-data-ui/releases/tag/v2.3.37) was published, with major type cleanup.

**Comment by hi-reeve** [comment]
i really hope a rewrite to use lang="ts" will happen. my problem is i want to use the slot legend and i want to create a component for custom legend. but currently i have no idea what the legend object type from the slot is

**Reply by graphieros** [reply]
@hi-reeve 

Thank you for pointing this out.
This can be achieved by improving the types in the provided .d.ts for now.
Currently, no slots are typed in component type definitions (`DefineComponent`), I'll be working on it.

**Reply by hi-reeve** [reply]
> @hi-reeve 
> 
> 
> 
> Thank you for pointing this out.
> 
> This can be achieved by improving the types in the provided .d.ts for now.
> 
> Currently, no slots are typed in component type definitions (`DefineComponent`), I'll be working on it.
> 
> 

thankyou!

**Reply by graphieros** [reply]
@hi-reeve 

You can upgrade to [3.18.0](https://github.com/graphieros/vue-data-ui/releases/tag/v3.18.0), with type defs for slots.
